File: freeling/suggestion.cc
#include <iostream>
#include <map>

#include "freeling.h"
#include "freeling/morfo/traces.h"

using namespace std;
using namespace freeling;

void ProcessSentences(list<sentence> &ls) {
  sentence::const_iterator previous_w;
  sentence::const_iterator human_w;
  int state = 0;
  wcout << L"(";
  for (list<sentence>::iterator is=ls.begin(); is!=ls.end(); is++) {
    for (sentence::const_iterator w=is->begin(); w!=is->end(); w++) {
      if (state == 0)
	{
	  if (w->get_form()==L"<" && w->get_lemma()==L"<")
	    {
	      previous_w = w;
	      previous_w--;
	      
	      state = 1;
	    }
	}
      if (state == 1)
	{
	  if (w->get_form()==L">" && w->get_lemma()==L">" && w->get_tag() == L"Fz")
	    {
	      human_w = w;
	      human_w--;
	      state = 2;
	    }
	}
     
      if (state == 2)
	{
	  if (w->get_form()==L">" && w->get_lemma()==L">")
	    {
	      wcout << L"(:human-suggestion \""<< human_w->get_form() << L"\" "; 
	      wcout << L":form \""<< previous_w->get_form() << L"\" :lemma \"" << previous_w->get_lemma() << L"\" :tag \"" << previous_w->get_tag() << L"\" " << L" :synset-list \""<< previous_w->get_senses_string() << "\")"; 
	      state = 0;
	    }
	}
    }
  }
  wcout << L")";
}




int main (int argc, char **argv) {
  util::init_locale(L"default");

  wstring ipath;
  if(argc < 2) ipath=L"/usr/local";
  else ipath=util::string2wstring(argv[1]);

  wstring lang=L"pt";
  wstring path=ipath+L"/share/freeling/"+lang+L"/";

  tokenizer tk(path+L"tokenizer.dat"); 
  splitter sp(path+L"splitter.dat");

  maco_options opt(L"es");  
  opt.UserMapFile=L"";
  opt.LocutionsFile=path+L"locucions.dat"; opt.AffixFile=path+L"afixos.dat";
  opt.ProbabilityFile=path+L"probabilitats.dat"; opt.DictionaryFile=path+L"dicc.src";
  opt.NPdataFile=path+L"np.dat"; opt.PunctuationFile=path+L"../common/punct.dat"; 

  maco morfo(opt); 
  morfo.set_active_options (false,// UserMap
                             true, // NumbersDetection,
                             true, //  PunctuationDetection,
                             true, //  DatesDetection,
                             true, //  DictionarySearch,
                             true, //  AffixAnalysis,
                             false, //  CompoundAnalysis,
                             true, //  RetokContractions,
                             true, //  MultiwordsDetection,
                             true, //  NERecognition,
                             false, //  QuantitiesDetection,
                             true);  //  ProbabilityAssignment

  hmm_tagger tagger(path+L"tagger.dat", true, FORCE_TAGGER); 

  senses sen(path+L"senses.dat");
  ukb wsd(path+L"ukb.dat");
  wstring line,text;
  while (getline(wcin,line)) 
    text = text + line + L"\n";

  list<word> lw = tk.tokenize(text);
  list<sentence> ls = sp.split(lw);   
  morfo.analyze(ls);
  tagger.analyze(ls);
  sen.analyze(ls);
  wsd.analyze(ls);
 
  ProcessSentences(ls);    

}

File: freeling/suggestion.py
#! /usr/bin/python3

### REQUIRES python 3 !!!!
## For example:
##     ./suggestion.py <test.txt >test_out.txt

import freeling
import sys

FREELINGDIR = "/usr/local";

DATA = FREELINGDIR+"/share/freeling/";
LANG="pt";

freeling.util_init_locale("default");

# create language analyzer
la=freeling.lang_ident(DATA+"common/lang_ident/ident.dat");

# create options set for maco analyzer. Default values are Ok, except for data files.
op= freeling.maco_options("pt");
op.set_data_files( "", 
                   DATA + "common/punct.dat",
                   DATA + LANG + "/dicc.src",
                   DATA + LANG + "/afixos.dat",
                   "",
                   DATA + LANG + "/locucions.dat", 
                   DATA + LANG + "/np.dat",
                   DATA + "common/quantities_default.dat",
                   DATA + LANG + "/probabilitats.dat");

# create analyzers
tk=freeling.tokenizer(DATA+LANG+"/tokenizer.dat");
sp=freeling.splitter(DATA+LANG+"/splitter.dat");
sid=sp.open_session();
mf=freeling.maco(op);

# activate mmorpho odules to be used in next call
mf.set_active_options(False, True, True, True,  # select which among created 
                      True, True, False, True,  # submodules are to be used. 
                      True, True, True, True ); # default: all created submodules are used

# create tagger, sense anotator, and parsers
tg=freeling.hmm_tagger(DATA+LANG+"/tagger.dat",True,2);
sen=freeling.senses(DATA+LANG+"/senses.dat");
parser= freeling.chart_parser(DATA+LANG+"/chunker/grammar-chunk.dat");
wsd = freeling.ukb(DATA+LANG+"/ukb.dat");
# dep=freeling.dep_txala(DATA+LANG+"/dep/dependences.dat", parser.get_start_symbol());

# process input text
lin=sys.stdin.readline();
print("(")
#XSprint ("Text language is: "+la.identify_language(lin,["es","ca","en","it"])+"\n");
state = 0;
while (lin) :
        
    l = tk.tokenize(lin);
    ls = sp.split(sid,l,False);

    ls = mf.analyze(ls);
    ls = tg.analyze(ls);
    ls = sen.analyze(ls);
    ls = parser.analyze(ls);
    ls = wsd.analyze(ls);
    # ls = dep.analyze(ls);
    tempo = 0
    ## output results
    for s in ls :
       ws = s.get_words();
       for w in ws :
           if(tempo == 0):
               human = w
               previous_helper = w
           tempo = tempo+1
           
           if (state == 0):
               if (w.get_form() == "<" and w.get_lemma()=="<" and w.get_tag() == "Fz"):
                   state = 1
           if (state == 1):
               if (w.get_form() == ">" and w.get_lemma()==">" and w.get_tag() == "Fz"):
                   state = 2
           if (state == 2):
               if (w.get_form() == ">" and w.get_lemma()==">"):
                   print("(:human-suggestion \""+human.get_form()+"\" :form \""+previous.get_form()+"\" :lemma \""+previous.get_lemma()+"\" :tag \""+previous.get_tag()+"\" :synset-list \""+previous.get_senses_string()+"\")")
                   state=0
                   print("")
           previous = previous_helper
           previous_helper = human
           human = w
 #          print(w.get_form()+" "+w.get_lemma()+" "+w.get_tag()+" "+w.get_senses_string());
      # print ("");
       
#       tr = s.get_parse_tree();
#       printTree(tr, 0);

 #      dp = s.get_dep_tree();
 #      printDepTree(dp, 0)

    lin=sys.stdin.readline();
print(")")    
# clean up       
sp.close_session(sid);
    
